pootle/apps/pootle_data/store_data.py

```diff
@@ -0,0 +1,160 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) Pootle contributors.
+#
+# This file is a part of the Pootle project. It is distributed under the GPL3
+# or later license. See the LICENSE file for a copy of the license and the
+# AUTHORS file for copyright and authorship information.
+
+from translate.filters.decorators import Category
+
+from django.db import models
+from django.db.models import Case, Count, Max, Q, When
+
+from pootle_statistics.models import SubmissionTypes
+from pootle_store.constants import FUZZY, OBSOLETE, TRANSLATED
+from pootle_store.models import QualityCheck
+from pootle_store.util import SuggestionStates
+
+from .utils import DataTool, DataUpdater
+
+
+class StoreDataUpdater(DataUpdater):
+
+    related_name = "store"
+    aggregate_fields = (
+        "words",
+        "last_update_revision",
+        "max_unit_revision",
+        "max_unit_mtime")
+
+    @property
+    def store(self):
+        return self.model
+
+    @property
+    def store_data_qs(self):
+        return self.store.unit_set
+
+    @property
+    def units(self):
+        """Non-obsolete units in this Store"""
+        return self.store_data_qs.filter(state__gt=OBSOLETE)
+
+    @property
+    def aggregate_words(self):
+        return dict(
+            total_words=models.Sum(
+                Case(
+                    When(Q(state__gt=OBSOLETE)
+                         & Q(source_wordcount__gt=0),
+                         then="source_wordcount"),
+                    default=0)),
+            translated_words=models.Sum(
+                Case(
+                    When(Q(state=TRANSLATED)
+                         & Q(source_wordcount__gt=0),
+                         then="source_wordcount"),
+                    default=0)),
+            fuzzy_words=models.Sum(
+                Case(
+                    When(Q(state=FUZZY)
+                         & Q(source_wordcount__gt=0),
+                         then="source_wordcount"),
+                    default=0)))
+
+    @property
+    def aggregate_last_update_revision(self):
+        return dict(
+            last_update_revision=Max(
+                Case(
+                    When(Q(state__gt=OBSOLETE)
+                         & Q(revision__isnull=False),
+                         then="revision"))))
+
+    @property
+    def aggregate_max_unit_revision(self):
+        return dict(max_unit_revision=Max("revision"))
+
+    @property
+    def aggregate_max_unit_mtime(self):
+        return dict(max_unit_mtime=Max("mtime"))
+
+    def get_last_updated_unit(self, **kwargs):
+        order_by = ("-revision", "-mtime", "-id")
+        if "last_update_revision" in kwargs:
+            if not kwargs["last_update_revision"]:
+                return
```

Comment:
this is here because if there is aggregate_data and there has been no revision found, there is no point looking for the last unit to be updated